execution.go

```diff
@@ -73,232 +73,165 @@ func NewEngineAPIExecutionClient(
+	firstTx := txs[0]
+	err = json.Unmarshal(firstTx, &payloadResult)
 	if err != nil {
-		return common.Hash{}, fmt.Errorf("failed to get block at height %d: %w", height, err)
-	}
-	// Extract the block hash
-	blockHashStr, ok := block["hash"].(string)
-	if !ok {
-		return common.Hash{}, fmt.Errorf("failed to get block hash at height %d", height)
+		return nil, 0, fmt.Errorf("failed to deserialize first transaction as ExecutionPayload: %w", err)
 	}
-	return common.HexToHash(blockHashStr), nil
-}
 
-// ExecuteTxs executes the given transactions and returns the new state root and gas used
-func (c *EngineAPIExecutionClient) ExecuteTxs(ctx context.Context, txs []execution_types.Tx, height uint64, timestamp time.Time, prevStateRoot execution_types.Hash) (execution_types.Hash, uint64, error) {
-	// convert rollkit tx to eth tx
-	ethTxs := make([]*types.Transaction, len(txs))
-	for i, tx := range txs {
-		ethTxs[i] = new(types.Transaction)
-		err := ethTxs[i].UnmarshalBinary(tx)
-		if err != nil {
-			return execution_types.Hash{}, 0, fmt.Errorf("failed to unmarshal transaction: %w", err)
-		}
+	// Add transactions from txs to the payload (skip the first one which is the payload itself)
+	payloadResult.ExecutionPayload.Transactions = make([][]byte, len(txs)-1)
+	for i := 1; i < len(txs); i++ {
+		payloadResult.ExecutionPayload.Transactions[i-1] = txs[i]
 	}
 
-	// encode
-	txsPayload := make([][]byte, len(txs))
-	for i, tx := range ethTxs {
-		buf := bytes.Buffer{}
-		err := tx.EncodeRLP(&buf)
-		if err != nil {
-			return execution_types.Hash{}, 0, fmt.Errorf("failed to RLP encode tx: %w", err)
-		}
-		txsPayload[i] = buf.Bytes()
+	var newPayloadResult engine.PayloadStatusV1
+	err = c.engineClient.CallContext(ctx, &newPayloadResult, "engine_newPayloadV3",
+		payloadResult.ExecutionPayload,
+		[]string{}, // No blob hashes
+		c.genesisHash.Hex(),
+	)
+
+	if err != nil {
+		return nil, 0, fmt.Errorf("new payload submission failed: %w", err)
 	}
 
-	// fetch previous block hash to update forkchoice for the next payload id
-	var err error
-	var prevBlockHash common.Hash
-	if height == c.initialHeight {
-		prevBlockHash = c.genesisHash
-	} else {
-		prevBlockHash, err = c.getBlockHash(ctx, height-1)
-		if err != nil {
-			return execution_types.Hash{}, 0, err
-		}
+	if newPayloadResult.Status != engine.VALID {
+		return nil, 0, fmt.Errorf("new payload submission failed with: %s", *newPayloadResult.ValidationError)
 	}
 
-	// update forkchoice to get the next payload id
+	// forkchoice update
+	blockHash := payloadResult.ExecutionPayload.BlockHash
 	var forkchoiceResult engine.ForkChoiceResponse
 	err = c.engineClient.CallContext(ctx, &forkchoiceResult, "engine_forkchoiceUpdatedV3",
 		engine.ForkchoiceStateV1{
-			HeadBlockHash:      prevBlockHash,
-			SafeBlockHash:      prevBlockHash,
-			FinalizedBlockHash: prevBlockHash,
+			HeadBlockHash: blockHash,
+			SafeBlockHash: blockHash,
+			// FinalizedBlockHash: blockHash,
```

Comment:
i guess this is on purpose to allow for SetFinal to set this?

Reply:
Exactly